support/sier-codec/src/schema.rs

```diff
@@ -0,0 +1,245 @@
+                    type_: Type::U8,
+                },
+            ],
+        };
+
+        let parsed = struct_def.parse(&[42, 43, 44]).unwrap();
+        assert_eq!(parsed["bar"], Value::U8(42));
+        assert_eq!(parsed["baz"], Value::U8(43));
+        assert_eq!(parsed["qux"], Value::U8(44));
+    }
+
+    #[test]
+    fn too_few_bytes() {
+        let struct_def = StructDef {
+            type_name: "Foo".to_string(),
+            fields: vec![FieldDef {
+                name: "bar".to_string(),
+                type_: Type::U32,
+            }],
+        };
+
+        let result = struct_def.parse(&[42, 0]);
+        assert!(result.is_err());
+    }
+
+    #[test]
+    fn too_many_bytes() {
+        let struct_def = StructDef {
+            type_name: "Foo".to_string(),
+            fields: vec![FieldDef {
+                name: "bar".to_string(),
+                type_: Type::U8,
+            }],
+        };
+
+        let result = struct_def.parse(&[42, 0]);
+        assert!(result.is_err());
+    }
+
+    #[cfg(test)]
+    mod id {
+        use super::*;
+
+        #[test]
+        fn is_different_with_different_field_names() {
+            let struct_a = StructDef {
+                type_name: "Foo".to_string(),
+                fields: vec![FieldDef {
+                    name: "bar".to_string(),
+                    type_: Type::U8,
+                }],
+            };
+            let struct_b = StructDef {
+                type_name: "Foo".to_string(),
+                fields: vec![FieldDef {
+                    name: "baz".to_string(),
+                    type_: Type::U8,
+                }],
+            };
+
+            assert_ne!(struct_a.id(), struct_b.id());
+        }
+
+        #[test]
+        fn is_different_with_different_field_types() {
+            let struct_a = StructDef {
+                type_name: "Foo".to_string(),
+                fields: vec![FieldDef {
+                    name: "bar".to_string(),
+                    type_: Type::U8,
+                }],
+            };
+            let struct_b = StructDef {
+                type_name: "Foo".to_string(),
+                fields: vec![FieldDef {
+                    name: "bar".to_string(),
+                    type_: Type::U32,
+                }],
+            };
+
+            assert_ne!(struct_a.id(), struct_b.id());
+        }
+
+        #[test]
+        fn is_same_with_same_fields() {
+            let struct_a = StructDef {
+                type_name: "Foo".to_string(),
+                fields: vec![FieldDef {
+                    name: "bar".to_string(),
+                    type_: Type::U8,
+                }],
+            };
+            let struct_b = StructDef {
+                type_name: "Foo".to_string(),
+                fields: vec![FieldDef {
+                    name: "bar".to_string(),
+                    type_: Type::U8,
+                }],
+            };
+
+            assert_eq!(struct_a.id(), struct_b.id());
+        }
```

Comment:
panic case on id is not covered, maybe because of logical reasons it can never occur?

Reply:
The expect should never panic. We know the digest output is always 64 bytes.

I don't see a combination of methods in generic_array (https://docs.rs/generic-array/0.14.4/generic_array/struct.GenericArray.html) and std to get a subarray of fixed size from a larger one. So we just use the easier methods that return Option<[u8; 8]> that we need to unwrap.